src/PowerShellEditorServices/Services/DebugAdapter/DebugService.cs

```diff
@@ -456,22 +457,25 @@ public async Task<string> SetVariableAsync(int variableContainerReferenceId, str
 
             if (argTypeConverterAttr is not null)
             {
+                // PSVariable *is* strongly typed, so we have to convert it.
                 _logger.LogTrace($"Setting variable '{name}' using conversion to value: {expressionResult ?? "<null>"}");
 
-                // TODO: This is throwing a 'PSInvalidOperationException' thus causing
-                // 'DebuggerSetsVariablesWithConversion' to fail.
-                psVariable.Value = await _executionService.ExecuteDelegateAsync(
-                    "PS debugger argument converter",
-                    ExecutionOptions.Default,
-                    (pwsh, _) =>
-                    {
-                        var engineIntrinsics = (EngineIntrinsics)pwsh.Runspace.SessionStateProxy.GetVariable("ExecutionContext");
-
-                        // TODO: This is almost (but not quite) the same as LanguagePrimitives.Convert(), which does not require the pipeline thread.
-                        //       We should investigate changing it.
-                        return argTypeConverterAttr.Transform(engineIntrinsics, expressionResult);
-                    },
+                // NOTE: We use 'Get-Variable' here instead of 'SessionStateProxy.GetVariable()'
+                // because we already have a pipeline running (the debugger) and the latter cannot
+                // run concurrently (threw 'NoSessionStateProxyWhenPipelineInProgress').
+                IReadOnlyList<EngineIntrinsics> results = await _executionService.ExecutePSCommandAsync<EngineIntrinsics>(
+                    new PSCommand()
+                        .AddCommand(@"Microsoft.PowerShell.Utility\Get-Variable")
```

Comment:
@rjmholt Do you have a better idea or explanation as to why that exception was being thrown?

Reply:
This looks correct to me.  The debugger knows it is running in an active pipeline and runs commands as 'nested' on the target (locally or remotely), whereas SessionStateProxy does not.